sys/xtimer/xtimer_core.c

```diff
@@ -320,251 +259,119 @@ void xtimer_remove(xtimer_t *timer)
-
 /**
- * @brief compare two timers' target values, return the one with lower value.
- *
- * if either is NULL, return the other.
- * if both are NULL, return NULL.
+ * @brief update long timers' offsets and switch those that will expire in
+ *        one short timer period to the short timer list
  */
-static inline xtimer_t *_compare(xtimer_t *a, xtimer_t *b)
+static inline void _update_long_timers(uint32_t *now)
 {
-    if (a && b) {
-        return ((a->target <= b->target) ? a : b);
-    }
-    else {
-        return (a ? a : b);
-    }
-}
+    if (long_list_head) {
+        xtimer_t *curr_timer = long_list_head;
 
-/**
- * @brief merge two timer lists, return head of new list
- */
-static xtimer_t *_merge_lists(xtimer_t *head_a, xtimer_t *head_b)
-{
-    xtimer_t *result_head = _compare(head_a, head_b);
-    xtimer_t *pos = result_head;
-
-    while (1) {
-        head_a = head_a->next;
-        head_b = head_b->next;
-        if (!head_a) {
-            pos->next = head_b;
-            break;
-        }
-        if (!head_b) {
-            pos->next = head_a;
-            break;
-        }
+        while (curr_timer) {
+            uint32_t elapsed = *now - curr_timer->start_time;
+            if (curr_timer->offset <= elapsed) {
+                curr_timer->long_offset--;
+            }
+            curr_timer->offset -= elapsed;
+            curr_timer->start_time = *now;
 
-        pos->next = _compare(head_a, head_b);
-        pos = pos->next;
-    }
+            if (!curr_timer->long_offset) {
+                assert(curr_timer == long_list_head);
 
-    return result_head;
+                xtimer_t *trans_timer = curr_timer;
+                curr_timer = curr_timer->next;
+                _remove_timer_from_list(&long_list_head, trans_timer);
+                _add_timer_to_list(&timer_list_head, trans_timer);
+            }
+            else {
+                curr_timer = curr_timer->next;
+            }
+        }
+    }
 }
 
 /**
- * @brief parse long timers list and copy those that will expire in the current
- *        short timer period
+ * @brief update short timers' offsets and fire those that are close to expiry
  */
-static void _select_long_timers(void)
-{
-    xtimer_t *select_list_start = long_list_head;
-    xtimer_t *select_list_last = NULL;
-
-    /* advance long_list head so it points to the first timer of the next (not
-     * just started) "long timer period" */
-    while (long_list_head) {
-        if ((long_list_head->long_target <= _long_cnt) && _this_high_period(long_list_head->target)) {
-            select_list_last = long_list_head;
-            long_list_head = long_list_head->next;
+static inline void _update_short_timers(uint32_t *now) {
+    while (timer_list_head) {
+        uint32_t elapsed = *now - timer_list_head->start_time;
+
+        if (timer_list_head->offset <= elapsed ||
+            timer_list_head->offset - elapsed < XTIMER_ISR_BACKOFF) {
+            /* make sure we don't fire too early */
```

Comment:
can you use `_xtimer_spin()` here?

Reply:
We can but I think it could be more inefficient. This is because _xtimer_now() should be called one more to calculate an offset (the input for _xtimer_spin() function). Executing _xtimer_now() takes non-negligible time depending on platforms.